src/bidiMapper/domains/context/bidiPreloadScript.ts

```diff
@@ -0,0 +1,140 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+import type {CommonDataTypes, Script} from '../../../protocol/protocol.js';
+import {uuidv4} from '../../../utils/uuid.js';
+import type {CdpTarget} from './cdpTarget.js';
+
+type CdpPreloadScript = {
+  /** CDP target. Includes session ID and target ID. */
+  target: CdpTarget;
+  /** CDP preload script ID. */
+  preloadScriptId: Script.PreloadScript;
+};
+
+/**
+ * BiDi IDs are generated by the server and are unique within the context.
+ *
+ * CDP preload script IDs are generated by the client and are unique
+ * within the session.
+ *
+ * The mapping between BiDi and CDP preload script IDs is 1:many.
+ * BiDi IDs are needed by the mapper to keep track of potential multiple CDP IDs
+ * in the client.
+ */
+export class BidiPreloadScript {
+  /** BiDi ID, an automatically generated UUID. */
+  readonly #id: string = uuidv4();
+  /** CDP preload scripts. */
+  #cdpPreloadScripts: CdpPreloadScript[] = [];
+  /** The script itself, in a format expected by the spec i.e. a function. */
+  readonly #functionDeclaration: string;
+  /** Browsing context ID. */
+  readonly #contextId: CommonDataTypes.BrowsingContext | null;
+  readonly #targetIds: Set<string> = new Set();
+
+  get id(): string {
+    return this.#id;
+  }
+
+  get contextId(): CommonDataTypes.BrowsingContext | null {
+    return this.#contextId;
+  }
+
+  get targetIds(): Set<string> {
+    return this.#targetIds;
+  }
+
+  constructor(params: Script.AddPreloadScriptParameters) {
+    if (params.sandbox !== undefined) {
+      // TODO: Handle sandbox.
+      throw new Error('Sandbox is not supported yet');
+    }
+    if (params.arguments !== undefined && params.arguments.length > 0) {
+      // TODO: Handle arguments.
+      throw new Error('add preload script arguments are not supported');
+    }
+
+    this.#functionDeclaration = params.functionDeclaration;
+    this.#contextId = params.context ?? null;
+  }
+
+  async initInTargets(cdpTargets: Set<CdpTarget>) {
+    await Promise.all(
+      Array.from(cdpTargets.values()).map((cdpTarget) =>
+        this.initInTarget(cdpTarget)
+      )
+    );
+  }
+
+  /**
+   * Adds the script to the given CDP target by
+   * `Page.addScriptToEvaluateOnNewDocument` command.
+   * @param cdpTarget
+   */
+  async initInTarget(cdpTarget: CdpTarget) {
+    const addCdpPreloadScriptResult = await cdpTarget.cdpClient.sendCommand(
+      'Page.addScriptToEvaluateOnNewDocument',
+      {
+        source: `(${this.#functionDeclaration})();`,
+      }
+    );
+
+    this.#cdpPreloadScripts.push({
+      target: cdpTarget,
+      preloadScriptId: addCdpPreloadScriptResult.identifier,
+    });
+    this.#targetIds.add(cdpTarget.targetId);
+  }
+
+  /**
+   * Schedules the script to be run right after
+   * `Runtime.runIfWaitingForDebugger`, but does not wait for result.
```

Comment:
Add:

`XXX: Make it possible to evaluate in a specific sandbox / worldName` (or similar)

Reply:
True, it's redundant, I think it would just be clearer in terms of TODO for us, to remember what specific part of CDP we want to extend.

I am fine if you prefer to omit it.